### PR TITLE
Automatically reconnecting to the database when safe to do so

### PR DESCRIPTION
**Pull request by lorenzo**
This has been anf often requested feature for the database layer,
specially by people using the database in long running background tasks.

Instead of instructing them to figure out how to configure their database
to not drop connections, better just give them an extra hand by reconnecting
to the database whenever it is safe to do so.

That is, reconnection are attempted only on fresh commands to the database
outside of a transaction.

### CODE REVIEW

File: src/Database/Retry/ReconnectStrategy.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+use Cake\Core\Retry\RetryStrategyInterface;
+use Cake\Database\Connection;
+use Exception;
+
+/**
+ * Makes sure the connection to the database is alive before authorizing
+ * the retry of an action.
+ */
+class ReconnectStrategy implements RetryStrategyInterface
```
Comment (markstory) [inline src/Database/Retry/ReconnectStrategy.php]: Should this class be marked as internal?

File: tests/TestCase/Database/ConnectionTest.php [hunk 2]
```diff
@@ -1215,4 +1219,66 @@ public function pushNestedTransactionState()
...
+            ->will($this->throwException(new Exception('server gone away')));
+
+        $newDriver->expects($this->at(1))->method('disconnect');
+        $newDriver->expects($this->at(2))->method('connect');
+        $newDriver->expects($this->at(3))
+            ->method('prepare')
+            ->will($this->returnValue($statement));
+
+        $res = $conn->query('SELECT 1');
+        $this->assertInstanceOf('Cake\Database\StatementInterface', $res);
```
Comment (markstory) [inline tests/TestCase/Database/ConnectionTest.php]: You could use `StatementInterface::class` here.

**Comment by markstory** [comment]
Nice work @lorenzo. I like the the decorator based approach.

File: src/Database/Retry/ReconnectStrategy.php (hunk 1 above)
Comment (lorenzo) [reply src/Database/Retry/ReconnectStrategy.php]: I’d like to eventually make this visible as a thing, for example it is useful for retrying http requests

Comment (lorenzo) [reply src/Database/Retry/ReconnectStrategy.php]: Well, not this class. The CommandRetry is what I was speaking about.

it is sensible to mark this one as internal, I agree

**Comment by dereuromark** [comment]
Refs 
- https://github.com/cakephp/cakephp/issues/3423
- https://github.com/cakephp/cakephp/issues/11650

**Comment by lorenzo** [comment]
There is an unrelated phpstan failing check, what should we do about it?

**Comment by ADmad** [comment]
@lorenzo Ignore it :).  I'll take care of it on `3.next` branch.

**Pull request merged**